include/seqan3/core/metafunction/range.hpp

```diff
@@ -0,0 +1,280 @@
+ * \tparam t The type you wish to query; must satisfy seqan3::input_range_concept.
+ */
+template <input_range_concept rng_t>
+//!\cond
+    requires !iterator_concept<rng_t>
+//!\endcond
+struct rvalue_reference<rng_t>
+{
+    //!\brief Return the rvalue_reference member definition from the queried type's iterator.
+    using type = rvalue_reference_t<iterator_t<rng_t>>;
+};
+
+// ----------------------------------------------------------------------------
+// const_reference
+// ----------------------------------------------------------------------------
+
+/*!\brief Type metafunction that returns the `const_reference` of another type [specialisation for input ranges].
+ * \tparam t The type you wish to query; must satisfy seqan3::input_range_concept.
+ */
+template <input_range_concept rng_t>
+//!\cond
+    requires !iterator_concept<rng_t>
+//!\endcond
+struct const_reference<rng_t>
+{
+    //!\brief Resolves to the reference type of the `const_iterator` of t (not the `const iterator`!).
+    using type = reference_t<iterator_t<rng_t const>>;
+};
+
+// ----------------------------------------------------------------------------
+// difference_type
+// ----------------------------------------------------------------------------
+
+/*!\brief Type metafunction that returns the `difference_type` of another type [specialisation for ranges].
+ * \tparam t The type you wish to query; must satisfy seqan3::input_range_concept.
+ */
+template <range_concept rng_t>
+//!\cond
+    requires !iterator_concept<rng_t>
+//!\endcond
+struct difference_type<rng_t>
+{
+    //!\brief Return the difference_type member definition from the queried type's iterator.
+    using type = difference_type_t<iterator_t<rng_t>>;
+};
+
+// ----------------------------------------------------------------------------
+// size_type
+// ----------------------------------------------------------------------------
+
+/*!\brief Type metafunction that returns the `size_type` of another type [specialisation for sized ranges].
+ * \tparam t The type you wish to query; must satisfy seqan3::sized_range_concept.
+ */
+template <sized_range_concept rng_t>
+//!\cond
+    requires !iterator_concept<rng_t>
+//!\endcond
+struct size_type<rng_t>
+{
+    //!\brief Return the size_type member definition from the queried type's iterator.
+    using type = size_type_t<iterator_t<rng_t>>;
+};
+
+// ----------------------------------------------------------------------------
+// innermost_value_type
+// ----------------------------------------------------------------------------
+
+/*!\brief Recursively determines the `value_type` on containers and/or iterators [Type metafunction].
+ * \tparam t The type to recurse on; must have `ranges::value_type_t<rng_t>`
+ *
+ * \details
+ *
+ * Attention, this metafunction implicitly removes cv-qualifiers on the all value_types except the one returned.
+ */
+template <typename t>
+//!\cond
+    requires requires (t) { typename value_type_t<remove_cvref_t<t>>; }
```

Comment:
This function works on iterators and ranges, but it is usually only used in the context of ranges. It is not _pre, because it's not a base template, and it's not basic, because it's not basic :wink: 

We could put it to iterator, too, but it doesn't really make any diference.